aws/cloudformation/drone-stack.yml

```diff
@@ -0,0 +1,367 @@
+Parameters:
+  VpcId:
+    Type: String
+    Description: VpcId
+  SubnetIds:
+    Type: CommaDelimitedList
+    Description: Comma separated list of existing VPC Subnet Ids where ECS instances will run
+  AZs:
+    Type: CommaDelimitedList
+    Description: Comma separated list of availability zones where drone master ECS instances will run
+  NumWorkers:
+    Type: Number
+    Description: Initial number of drone workers (also the number of concurrent builds which can run)
+    Default: 2
+Resources:
+# ELB resources:
+# ELB / Target Group / Listener / ELB Security Group
+
+  ElbSecurityGroup:
+    Type: AWS::EC2::SecurityGroup
+    Properties:
+      GroupDescription: ELB Allowed Ports
+      VpcId: !Ref VpcId
+      SecurityGroupIngress:
+        - IpProtocol: tcp
+          FromPort: 80
+          ToPort: 80
+          CidrIp: 0.0.0.0/0
+      Tags:
+        -
+          Key: Name
+          Value: !Sub "ECS ${AWS::StackName} - ALB SecurityGroup"
+
+  DroneMasterTargetGroup:
+    Type: AWS::ElasticLoadBalancingV2::TargetGroup
+    Properties:
+      Name: !Sub "TargetGroup-${AWS::StackName}"
+      VpcId: !Ref VpcId
+      Port: 80
+      Protocol: HTTP
+      Tags:
+        -
+          Key: Name
+          Value: !Sub "ECS ${AWS::StackName} - TargetGroup"
+
+  DroneMasterELB:
+    Type: AWS::ElasticLoadBalancingV2::LoadBalancer
+    Properties:
+      Name: !Sub "ELB-${AWS::StackName}"
+      SecurityGroups:
+        - !Ref ElbSecurityGroup
+      Subnets: !Ref SubnetIds
+      Tags:
+        -
+          Key: Name
+          Value: !Sub "ECS ${AWS::StackName} - ELB"
+
+  LoadBalancerListener:
+    Type: AWS::ElasticLoadBalancingV2::Listener
+    Properties:
+      LoadBalancerArn: !Ref DroneMasterELB
+      Port: 80
+      Protocol: HTTP
+      DefaultActions:
+        - Type: forward
+          TargetGroupArn: !Ref DroneMasterTargetGroup
+
+# Drone Master resources:
+# Security Group / EC2 Launch Template / EC2 Autoscaling Group
+# ECS cluster / ECS service / ECS task definition
+
+  DroneMasterEcsSecurityGroup:
+    Type: AWS::EC2::SecurityGroup
+    Properties:
+      GroupDescription: ECS Allowed Ports
+      VpcId: !Ref VpcId
+      SecurityGroupIngress:
+        -
+          IpProtocol: tcp
+          FromPort: 1
+          ToPort: 65535
+          SourceSecurityGroupId: !Ref ElbSecurityGroup
+        -
+          IpProtocol: tcp
+          FromPort: 22
+          ToPort: 22
+          CidrIp: 0.0.0.0/0
+
+  DroneMasterECSCluster:
+    Type: AWS::ECS::Cluster
+
+  DroneMasterLaunchTemplate:
+    Type: AWS::EC2::LaunchTemplate
+    Properties:
+      LaunchTemplateName: !Sub "DroneMasterLaunchTemplate-${AWS::StackName}"
+      LaunchTemplateData: 
+        IamInstanceProfile: 
+          Name: ecsInstanceRole
+        KeyName: !Sub "winter-dev-${AWS::Region}"
+        SecurityGroupIds: 
+          - !Ref DroneMasterEcsSecurityGroup
+        ImageId: ami-0c4b05c0409735671
```

Comment:
You can use a reference to a [public parameter for ECS-Optimized AMI](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/retrieve-ecs-optimized_AMI.html) to make this more descriptive / region-agnostic. Two options:

1. [dynamic reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html) to a fixed value: `{{resolve:ssm:/aws/service/ecs/optimized-ami/amazon-linux/amzn-ami-2018.03.o-amazon-ecs-optimized/image_id:1}}`
The limitation that "You cannot currently specify that AWS CloudFormation use the latest version of a parameter" means that you have to put a specific version in the template.

2. Parameter reference to the latest recommended value (example from the documentation):

```
# Add to `Parameters` section:
  ECSAMI:
    Description: AMI ID
    Type: AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>
    Default: /aws/service/ecs/optimized-ami/amazon-linux/recommended/image_id

# Use the reference:
    ImageId: !Ref ECSAMI
```

Reply:
Thanks! I'll try option 2.